src/js/crypto.cpp

```diff
@@ -660,6 +660,130 @@ namespace ccf::js
     }
   }
 
+  static JSValue js_unwrap_key(
+    JSContext* ctx, JSValueConst, int argc, JSValueConst* argv)
+  {
+    if (argc != 3)
+      return JS_ThrowTypeError(
+        ctx, "Passed %d arguments, but expected 3", argc);
+
+    // API loosely modeled after
+    // https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey.
+
+    size_t key_size;
+    uint8_t* key = JS_GetArrayBuffer(ctx, &key_size, argv[0]);
+    if (!key)
+    {
+      return ccf::js::constants::Exception;
+    }
+
+    size_t unwrapping_key_size;
+    uint8_t* unwrapping_key =
+      JS_GetArrayBuffer(ctx, &unwrapping_key_size, argv[1]);
+    if (!unwrapping_key)
+    {
+      return ccf::js::constants::Exception;
+    }
+
+    js::Context& jsctx = *(js::Context*)JS_GetContextOpaque(ctx);
+
+    auto parameters = argv[2];
+    auto wrap_algo_name_val = jsctx(JS_GetPropertyStr(ctx, parameters, "name"));
+    JS_CHECK_EXC(wrap_algo_name_val);
+
+    auto wrap_algo_name_str = jsctx.to_str(wrap_algo_name_val);
+    if (!wrap_algo_name_str)
+    {
+      return ccf::js::constants::Exception;
+    }
+
+    try
+    {
+      auto algo_name = std::string(*wrap_algo_name_str);
+      if (algo_name == "RSA-OAEP")
+      {
+        // key can in principle be arbitrary data (see note on maximum size
+        // in rsa_key_pair.h). unwrapping_key is a private RSA key.
+
+        auto label_val = jsctx(JS_GetPropertyStr(ctx, parameters, "label"));
+        JS_CHECK_EXC(label_val);
+
+        size_t label_buf_size = 0;
+        uint8_t* label_buf = JS_GetArrayBuffer(ctx, &label_buf_size, label_val);
+
+        std::optional<std::vector<uint8_t>> label_opt = std::nullopt;
+        if (label_buf && label_buf_size > 0)
+        {
+          label_opt = {label_buf, label_buf + label_buf_size};
+        }
+
+        auto unwrapped_key = crypto::ckm_rsa_pkcs_oaep_unwrap(
+          crypto::Pem(unwrapping_key, unwrapping_key_size),
```

Comment:
would be best to construct outside, and call OPENSSL_cleanse() on, to avoid having another potential copy somewhere on the heap.